libs/agno/agno/knowledge/knowledge.py

```diff
@@ -1447,57 +1432,68 @@ async def async_search(
+        if self.contents_db is None:
+            log_warning("No contents db provided. This is required for filtering.")
+            return set()
+        contents, _ = await self.aget_content()
+        valid_filters: Set[str] = set()
+        for content in contents:
+            if content.metadata:
+                valid_filters.update(content.metadata.keys())
+
+        return valid_filters
 
+    def _validate_filters(
+        self, filters: Optional[Dict[str, Any]], valid_metadata_filters: Set[str]
+    ) -> Tuple[Dict[str, Any], List[str]]:
         if not filters:
             return {}, []
 
         valid_filters: Dict[str, Any] = {}
         invalid_keys = []
 
         # If no metadata filters tracked yet, all keys are considered invalid
-        if self.valid_metadata_filters is None:
+        if valid_metadata_filters is None or not valid_metadata_filters:
             invalid_keys = list(filters.keys())
-            log_debug(f"No valid metadata filters tracked yet. All filter keys considered invalid: {invalid_keys}")
+            log_warning(f"No valid metadata filters tracked yet. All filter keys considered invalid: {invalid_keys}")
             return {}, invalid_keys
 
         for key, value in filters.items():
             # Handle both normal keys and prefixed keys like meta_data.key
             base_key = key.split(".")[-1] if "." in key else key
-            if base_key in self.valid_metadata_filters or key in self.valid_metadata_filters:
+            if base_key in valid_metadata_filters or key in valid_metadata_filters:
                 valid_filters[key] = value
             else:
                 invalid_keys.append(key)
-                log_debug(f"Invalid filter key: {key} - not present in knowledge base")
+                log_warning(f"Invalid filter key: {key} - not present in knowledge base")
 
         return valid_filters, invalid_keys
 
-    def add_filters(self, metadata: Dict[str, Any]) -> None:
-        if self.valid_metadata_filters is None:
-            self.valid_metadata_filters = set()
+    def validate_filters(self, filters: Optional[Dict[str, Any]]) -> Tuple[Dict[str, Any], List[str]]:
+        valid_filters_from_db = self.get_valid_filters()
 
-        if metadata is not None:
-            for key in metadata.keys():
-                self.valid_metadata_filters.add(key)
+        valid_filters, invalid_keys = self._validate_filters(filters, valid_filters_from_db)
 
-    @cached_property
-    def _get_filters_from_db(self) -> Set[str]:
-        if self.contents_db is None:
-            return set()
-        contents, _ = self.get_content()
-        valid_filters: Set[str] = set()
-        for content in contents:
-            if content.metadata:
-                valid_filters.update(content.metadata.keys())
-        return valid_filters
+        return valid_filters, invalid_keys
+
+    async def async_validate_filters(self, filters: Optional[Dict[str, Any]]) -> Tuple[Dict[str, Any], List[str]]:
```

Comment:
```suggestion
    async def async_validate_filters(self, filters: Optional[Dict[str, Any]]) -> Tuple[Dict[str, Any], List[str]]:
        """Return a tuple containing a dict with all valid filters and a list of invalid filter keys"""
```

docstring as it's public